doc/tut3.rst

```diff
@@ -0,0 +1,277 @@
+for the compiler.  The way this works in Nim is, whenever the compiler
+encounters a call expression to a macro, The compiler evaluates the
+macro at compile time with the syntax tree from the invocation, and
+then it replacing the call to the macro by the result of the macro.
+
+Examples of things that can be implemented in macros:
+
+ * An assert macro that prints both sides of a comparison operator, if
+the assertion fails. ``myAssert(a == b)`` that converts to
+``if a != b: quit($a " != " $b)``
+
+ * A debug macro that prints the value and the name of the symbol.
+``myDebugEcho(a)`` that converts to ``echo "a: ", a``
+
+ * Symbolic differenciation of an expression.
+``diff(a*pow(x,3) + b*pow(x,2) + c*x + d, x)``  that converts to
+``3*a*pow(x,2) + 2*a*x + c``
+
+Macro Arguments
+---------------
+
+There are two ways to pass arguments to a macro, either typed or
+untyped. And the argument types have two faces.  One face is used for
+the overload resolution, and the other face is used for the semantic
+checking of the macro implementation.  For example
+``macro foo(arg: int)`` will be called in an expression ``foo(x)``, if
+``x`` is of type int, but for the semantic checking of the macro
+itself, ``arg`` has the type ``NimNode``, not ``int`` as you might
+expect, because ``x`` will be passed as a symbol (``NimNode``), not as
+an integer.
+
+Untyped Arguments
+-----------------
+
+Untyped macro arguments are passed to the macro before they are
+semantically checked. This means the syntax tree that is passed down
+to the macro does not need to make sense for Nim yet, the only
+limitation for the syntax in an untyped macro argument is, it needs to
+be parseable by the Nim parser.  The semantic of this syntax is
+entirely up to the macro implementation. In this case the macro is
+responsible to implement its own semantic checking on the
+argument. The upside for untyped arguments is, the syntax tree is
+quite predictable and less complex than for typed arguments.  Untyped
+arguments have the type ``untyped`` in arguments list.
+
+Typed Arguments
+---------------
+
+For typed arguments, the semantic checker runs on the argument and
+does transformations on it, before it is passed to the macro. Here
+identifier nodes will already be resolved as symbols, implicit type
+conversions are visible in the tree as calls, templates will be
+expanded and probably most importantly, nodes have type information.
+Typed arguments can have the type ``typed`` in the arguments list.
+But all other types, such as ``int``, ``float`` or ``MyObjectType``
+are typed arguments as well, and they will be passed to the macro as a
+syntax tree.
+
+Static Arguments
+----------------
+
+Static arguments are a way to pass values as values and not as syntax
+tree nodes to a macro. For example for ``macro foo(arg: static[int])``
+in the expression ``foo(x)``, ``x`` needs to be an integer constant,
+but in the macro body ``arg`` is just like a normal parameter of type
+``int``.
```

Comment:
A short example, please. (Something like it is done in 'Code blocks as arguments' section)